File: DSAA Interactive Dashboard/p1.c
#include <stdio.h>
#include <stdlib.h>

// ANSI Color Codes for high-tech look
#define RESET   "\033[0m"
#define CYAN    "\033[1;36m"
#define GREEN   "\033[1;32m"
#define RED     "\033[1;31m"
#define AMBER   "\033[1;33m"
#define PURPLE  "\033[1;35m"

#define MAX 100
int a[MAX], n = 0, total;

void printHeader(char* title) {
    printf(PURPLE "\n==================================================\n");
    printf("  [#] %s\n", title);
    printf("==================================================\n" RESET);
}

// FIXED: Cleaned buffer catch to wait perfectly for keypress
void printUserGuide() {
    printf(CYAN "\n==================================================\n");
    printf("         CRASH COURSE: HOW TO USE THIS SYSTEM\n");
    printf("==================================================\n" RESET);
    printf(" Think of this program as a digital shelf with numbered slots.\n\n");
    printf(AMBER " [STEP 1]" RESET " Define how many total slots you want to use today.\n");
    printf(AMBER " [STEP 2]" RESET " Use the menu numbers below to manage your data:\n");
    printf("   • " GREEN "Option 1 (Insert):" RESET " Type a value, then pick a slot (0, 1, 2...).\n");
    printf("   • " GREEN "Option 6 (View):" RESET "   Always use this to see your current shelf map.\n");
    printf("   • " GREEN "Option 2 (Delete):" RESET " Remove an item by typing its slot index number.\n");
    printf("   • " GREEN "Option 5 (Sort):" RESET "   Organizes scrambled numbers from lowest to highest.\n");
    printf("   • " GREEN "Option 3/4 (Search):" RESET "Finds what slot index a number is hiding in.\n");
    printf(PURPLE "==================================================\n\n" RESET);
    printf("Press " AMBER "[ENTER]" RESET " to initialize system config...");
    
    int ch;
    while ((ch = getchar()) != '\n' && ch != EOF); // Robust buffer purge
}

int is_empty() {
    if (n <= 0) { 
        printf(RED "\n [!] UNDERFLOW: Operational array contains zero elements.\n" RESET); 
        return 1; 
    }
    return 0;
}

// Simple internal helper to check if array layout is sorted
int is_sorted() {
    for (int i = 0; i < n - 1; i++) {
        if (a[i] > a[i+1]) return 0;
    }
    return 1;
}

void insert() {
    int v, p, i;
    if (n >= MAX || n >= total) { 
        printf(RED "\n [!] OVERFLOW: Allocation boundary limit reached.\n" RESET); 
        return; 
    }
    printf(AMBER " [?] Enter Value: " RESET); scanf("%d", &v);
    printf(AMBER " [?] Enter Position (0 to %d): " RESET, n); scanf("%d", &p);

    if (p < 0 || p > n) { 
        printf(RED " [!] ERROR: Target index falls outside variable parameters.\n" RESET); 
        return; 
    }

    for (i = n++; i > p; i--) a[i] = a[i-1];
    a[p] = v;
    printf(GREEN "  (+) Success: Element safely committed to index register.\n" RESET);
}

void delete_item() {
    int p, i;
    if (is_empty()) return;
    printf(AMBER " [?] Index to Delete (0 to %d): " RESET, n - 1); scanf("%d", &p);

    if (p < 0 || p >= n) { 
        printf(RED " [!] ERROR: Target index falls outside variable parameters.\n" RESET); 
        return; 
    }

    int removed = a[p];
    for (i = p; i < --n; i++) a[i] = a[i+1];
    printf(RED "  (-) Dropped: Memory cell released -> element [%d]\n" RESET, removed);
}

void search(int mode) {
    if (is_empty()) return;
    int v, i, found = -1;
    printf(AMBER " [?] Search Value: " RESET); scanf("%d", &v);

    if (mode == 1) { // Linear Search
        for (i = 0; i < n; i++) if (a[i] == v) { found = i; break; }
    } else { // Binary Search
        // FIXED: Safeguard check against unsorted query structures
        if (!is_sorted()) {
            printf(AMBER " [!] NOTICE: Memory block unsorted. Initializing fallback logic...\n" RESET);
            // Fallback to linear search so the user doesn't get a corrupt "Not Found" error
            for (i = 0; i < n; i++) if (a[i] == v) { found = i; break; }
        } else {
            int low = 0, high = n - 1;
            while (low <= high) {
                int mid = low + (high - low) / 2; // FIXED: Prevent overflow architecture bugs
                if (a[mid] == v) { found = mid; break; }
                (a[mid] < v) ? (low = mid + 1) : (high = mid - 1);
            }
        }
    }

    if (found != -1) printf(GREEN "  (*) Located: Element matching query found at index [%d]\n" RESET, found);
    else printf(RED " [!] ERROR: Query target match missing from active registers.\n" RESET);
}

void bubble_sort() {
    if (is_empty()) return;
    int i, j, t;
    for (i = 0; i < n - 1; i++)
        for (j = 0; j < n - i - 1; j++)
            if (a[j] > a[j+1]) { t = a[j]; a[j] = a[j+1]; a[j+1] = t; }
    printf(GREEN "  (^) Sort Complete: Active registry block sequences reordered.\n" RESET);
}

void display() {
    printHeader("STATIC DATA ARRAY VISUALIZATION");
    if (n == 0) {
        printf(RED "  [ !! SYSTEM EMPTY REGISTER FLUSH !! ] \n" RESET);
        return;
    }
    printf(CYAN " INDEX : " RESET);
    for (int i = 0; i < n; i++) printf("%-5d", i);
    printf(CYAN "\n DATA  : " RESET);
    for (int i = 0; i < n; i++) printf("[%d]  ", a[i]);
    printf("\n");
}

int main() {
    int choice;
    #ifdef _WIN32
        system("cls");
    #else
        system("clear");
    #endif

    printUserGuide();

    printf(CYAN "==================================================\n");
    printf("     [M] STATIC ARRAY CONTROL INTERFACE [M]\n");
    printf("==================================================\n" RESET);
    
    printf(AMBER " [?] Define Max Target Capacity: " RESET);
    scanf("%d", &total);

    while (1) {
        printf(CYAN "\n [SYSTEM STATUS] Active Blocks: %d/%d entries utilized\n" RESET, n, total);
        printf("  1.  (+) Insert Element\n");
        printf("  2.  (-) Delete Element\n");
        printf("  3.  (*) Run Linear Search query\n");
        printf("  4.  (*) Run Binary Search query\n");
        printf("  5.  (^) Run Structural Bubble Sort\n");
        printf("  6.  (#) Traverse Structural Visual\n");
        printf("  7.  (X) Terminate Control Kernel Execution\n");
        printf(AMBER " >>> Select Target Action Option: " RESET);
        
        if (scanf("%d", &choice) != 1) {
            // Clear input stream if strings/characters accidentally get passed
            while (getchar() != '\n');
            printf(RED " [!] ERROR: Invalid numeric token input.\n" RESET);
            continue;
        }

        switch (choice) {
            case 1: insert(); break;
            case 2: delete_item(); break;
            case 3: search(1); break;
            case 4: search(2); break;
            case 5: bubble_sort(); break;
            case 6: display(); break;
            case 7: 
                printf(RED "\n [!] CORE KERNEL POWER DOWN INTERRUPT ISSUED...\n" RESET);
                return 0;
            default: printf(RED " [!] WARNING: Operational instruction unrecognized.\n" RESET);
        }
    }
    return 0;
}

File: DSAA Interactive Dashboard/p1.py
import os
import sys

# ANSI Color Codes
RESET = "\033[0m"
CYAN = "\033[1;36m"
GREEN = "\033[1;32m"
RED = "\033[1;31m"
AMBER = "\033[1;33m"
PURPLE = "\033[1;35m"

MAX = 100
a = [0] * MAX
n = 0
total = 0

void = None # Simply a visual marker for matching style

def print_header(title):
    print(f"{PURPLE}\n==================================================")
    print(f"  [#] {title}")
    print(f"=================================================={RESET}")

def print_user_guide():
    print(f"{CYAN}\n==================================================")
    print("         CRASH COURSE: HOW TO USE THIS SYSTEM")
    print(f"=================================================={RESET}")
    print(" Think of this program as a digital shelf with numbered slots.\n")
    print(f"{AMBER} [STEP 1]{RESET} Define how many total slots you want to use today.")
    print(f"{AMBER} [STEP 2]{RESET} Use the menu numbers below to manage your data:")
    print(f"   • {GREEN}Option 1 (Insert):{RESET} Type a value, then pick a slot (0, 1, 2...).")
    print(f"   • {GREEN}Option 6 (View):{RESET}   Always use this to see your current shelf map.")
    print(f"   • {GREEN}Option 2 (Delete):{RESET} Remove an item by typing its slot index number.")
    print(f"   • {GREEN}Option 5 (Sort):{RESET}   Organizes scrambled numbers from lowest to highest.")
    print(f"   • {GREEN}Option 3/4 (Search):{RESET}Finds what slot index a number is hiding in.")
    print(f"{PURPLE}==================================================\n{RESET}")
    input(f"Press {AMBER}[ENTER]{RESET} to initialize system config...")

def is_empty():
    global n
    if n <= 0:
        print(f"{RED}\n [!] UNDERFLOW: Operational array contains zero elements.\n{RESET}")
        return True
    return False

def insert():
    global n, total, a
    if n >= MAX or n >= total:
        print(f"{RED}\n [!] OVERFLOW: Allocation boundary limit reached.\n{RESET}")
        return
    try:
        v = int(input(f"{AMBER} [?] Enter Value: {RESET}"))
        p = int(input(f"{AMBER} [?] Enter Position (0 to {n}): {RESET}"))
    except ValueError:
        print(f"{RED} [!] ERROR: Invalid numeric input.\n{RESET}")
        return

    if p < 0 or p > n:
        print(f"{RED} [!] ERROR: Target index falls outside variable parameters.\n{RESET}")
        return

    # Shift elements
    for i in range(n, p, -1):
        a[i] = a[i-1]
    a[p] = v
    n += 1
    print(f"{GREEN}  (+) Success: Element safely committed to index register.\n{RESET}")

def delete_item():
    global n, a
    if is_empty():
        return
    try:
        p = int(input(f"{AMBER} [?] Index to Delete (0 to {n - 1}): {RESET}"))
    except ValueError:
        print(f"{RED} [!] ERROR: Invalid numeric input.\n{RESET}")
        return

    if p < 0 or p >= n:
        print(f"{RED} [!] ERROR: Target index falls outside variable parameters.\n{RESET}")
        return

    removed = a[p]
    for i in range(p, n - 1):
        a[i] = a[i+1]
    n -= 1
    print(f"{RED}  (-) Dropped: Memory cell released -> element [{removed}]\n{RESET}")

def search(mode):
    global n, a
    if is_empty():
        return
    try:
        v = int(input(f"{AMBER} [?] Search Value: {RESET}"))
    except ValueError:
        print(f"{RED} [!] ERROR: Invalid numeric input.\n{RESET}")
        return
        
    found = -1
    if mode == 1: # Linear
        for i in range(n):
            if a[i] == v:
                found = i
                break
    else: # Binary
        low = 0
        high = n - 1
        while low <= high:
            mid = (low + high) // 2
            if a[mid] == v:
                found = mid
                break
            if a[mid] < v:
                low = mid + 1
            else:
                high = mid - 1

    if found != -1:
        print(f"{GREEN}  (*) Located: Element matching query found at index [{found}]\n{RESET}")
    else:
        print(f"{RED} [!] ERROR: Query target match missing from active registers.\n{RESET}")

def bubble_sort():
    global n, a
    if is_empty():
        return
    for i in range(n - 1):
        for j in range(n - i - 1):
            if a[j] > a[j+1]:
                a[j], a[j+1] = a[j+1], a[j]
    print(f"{GREEN}  (^) Sort Complete: Active registry block sequences reordered.\n{RESET}")

def display():
    global n, a
    print_header("STATIC DATA ARRAY VISUALIZATION")
    if n == 0:
        print(f"{RED}  [ !! SYSTEM EMPTY REGISTER FLUSH !! ] \n{RESET}")
        return
    print(f"{CYAN} INDEX : {RESET}", end="")
    for i in range(n):
        print(f"{i:<5}", end="")
    print(f"\n{CYAN} DATA  : {RESET}", end="")
    for i in range(n):
        print(f"[{a[i]}]  ", end="")
    print()

def main():
    global total
    os.system('cls' if os.name == 'nt' else 'clear')
    
    print_user_guide()

    print(f"{CYAN}==================================================")
    print("     [M] STATIC ARRAY CONTROL INTERFACE [M]")
    print(f"=================================================={RESET}")
    
    try:
        total = int(input(f"{AMBER} [?] Define Max Target Capacity: {RESET}"))
    except ValueError:
        print(f"{RED}Critical configuration failure. Exiting.{RESET}")
        return

    while True:
        print(f"{CYAN}\n [SYSTEM STATUS] Active Blocks: {n}/{total} entries utilized{RESET}")
        print("  1.  (+) Insert Element")
        print("  2.  (-) Delete Element")
        print("  3.  (*) Run Linear Search query")
        print("  4.  (*) Run Binary Search query")
        print("  5.  (^) Run Structural Bubble Sort")
        print("  6.  (#) Traverse Structural Visual")
        print("  7.  (X) Terminate Control Kernel Execution")
        
        choice = input(f"{AMBER} >>> Select Target Action Option: {RESET}")
        
        if choice == '1': insert()
        elif choice == '2': delete_item()
        elif choice == '3': search(1)
        elif choice == '4': search(2)
        elif choice == '5': bubble_sort()
        elif choice == '6': display()
        elif choice == '7':
            print(f"{RED}\n [!] CORE KERNEL POWER DOWN INTERRUPT ISSUED...\n{RESET}")
            sys.exit(0)
        else:
            print(f"{RED} [!] WARNING: Operational instruction unrecognized.\n{RESET}")

if __name__ == "__main__":
    main()